pkg/ccl/changefeedccl/cdctest/mock_webhook_sink.go

```diff
@@ -181,22 +243,76 @@ func (s *MockWebhookSink) requestHandler(hw http.ResponseWriter, hr *http.Reques
 
 func (s *MockWebhookSink) publish(hw http.ResponseWriter, hr *http.Request) error {
 	defer hr.Body.Close()
-	row, err := io.ReadAll(hr.Body)
-	if err != nil {
-		return err
-	}
+
 	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.mu.lastHeaders = hr.Header.Clone()
 	s.mu.numCalls++
-	if s.mu.statusCodes[s.mu.statusCodesIndex] >= http.StatusOK && s.mu.statusCodes[s.mu.statusCodesIndex] < http.StatusMultipleChoices {
-		s.mu.rows = append(s.mu.rows, string(row))
-		if s.mu.notify != nil {
-			close(s.mu.notify)
-			s.mu.notify = nil
+	statusCode := s.mu.statusCodes[s.mu.statusCodesIndex]
+	resBody, hasResBody := s.mu.responseBodies[s.mu.statusCodesIndex]
+	s.mu.statusCodesIndex = (s.mu.statusCodesIndex + 1) % len(s.mu.statusCodes)
+	s.mu.Unlock()
+
+	gzCompression := hr.Header.Get("Content-Encoding") == "gzip"
+
+	if statusCode < http.StatusOK || statusCode > http.StatusMultipleChoices {
+		if !hasResBody {
+			hw.WriteHeader(statusCode)
+			return nil
+		}
+
+		if gzCompression {
+			hw.Header().Set("Content-Encoding", "gzip")
+			gw := getEncoder(hw)
+			defer putEncoder(gw)
+			hw.WriteHeader(statusCode)
+			if hasResBody {
+				if _, err := gw.Write(resBody); err != nil {
+					return errors.Wrap(err, "failed to write response body")
+				}
+				if err := gw.Close(); err != nil {
+					return errors.Wrap(err, "failed to flush gzip writer")
+				}
+			}
+			return nil
+		}
+	}
+
+	var row []byte
+	if gzCompression {
+		gzReader, err := getDecoder(hr.Body)
+		if err != nil {
+			return errors.Wrap(err, "failed to create gzip reader")
+		}
+		defer putDecoder(gzReader)
+		row, err = io.ReadAll(gzReader)
+		if err != nil {
+			return errors.Wrap(err, "failed to read compressed request body")
+		}
+		if err = gzReader.Close(); err != nil {
+			return errors.Wrap(err, "failed to close gzip reader")
+		}
+	} else {
+		var err error
+		row, err = io.ReadAll(hr.Body)
+		if err != nil {
+			return errors.Wrap(err, "failed to read plain request body")
+		}
+	}
+
+	s.mu.Lock()
+	s.mu.rows = append(s.mu.rows, string(row))
+	if s.mu.notify != nil {
+		close(s.mu.notify)
+		s.mu.notify = nil
+	}
+	s.mu.Unlock()
+
+	hw.WriteHeader(statusCode)
+	if hasResBody {
+		if _, err := hw.Write(resBody); err != nil {
```

Comment:
Should we be using the encoder to write the body if compression is set like in the non-2XX code case?

Reply:
The webhook sink mock doesn't support response body compression for successful (2xx) responses. This is intentional since we only need to read and log response bodies when handling HTTP errors - in those cases, we want to capture all possible details, including the decoded body, for debugging purposes.